### PR TITLE
[FLINK-2004] Fix memory leak in presence of failed checkpoints for Kafka Source

### PR DESCRIPTION
**Pull request by rmetzger**

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/persistent/PersistentKafkaSource.java [hunk 1]
```diff
@@ -236,13 +246,22 @@ public void restoreState(long[] state) {
...
-		long[] checkpointOffsets = pendingCheckpoints.remove(checkpointId);
-		if(checkpointOffsets == null) {
+		final int posInMap = pendingCheckpoints.indexOf(checkpointId);
+		if(posInMap == -1) {
 			LOG.warn("Unable to find pending checkpoint for id {}", checkpointId);
 			return;
 		}
+
+		long[] checkpointOffsets = (long[]) pendingCheckpoints.remove(posInMap);
 		LOG.info("Got corresponding offsets {}", Arrays.toString(checkpointOffsets));
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/persistent/PersistentKafkaSource.java]: I think this log message can be improved. To something like "Committing offsets {} to Kafka."

**Comment by StephanEwen** [comment]
Looks good. Some of the tests look like unit-style tests that are sort of lightweight.

May be good to pull those out in to a `PersistentKafkaSourceTest`-

Otherwise, +1 to merge

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/persistent/PersistentKafkaSource.java [hunk 2]
```diff
@@ -225,7 +226,16 @@ public void close() {
 
 	@Override
 	public void restoreState(long[] state) {
-		// we maintain the offsets in Kafka, so nothing to do.
+		if(lastOffsets == null) {
+			LOG.warn("Restore state called before open() has been called");
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/persistent/PersistentKafkaSource.java]: Actually, restore state is always called before `open()`, as far as I know.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/persistent/PersistentKafkaSource.java [hunk 3]
```diff
@@ -225,7 +226,16 @@ public void close() {
...
+			return;
+		}
+		LOG.info("Restoring state to {}", Arrays.toString(state));
+		if(lastOffsets.length != state.length) {
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/persistent/PersistentKafkaSource.java]: How about sanity checking before logging that things are going to happen? Usually gives better logging insights...

**Comment by StephanEwen** [comment]
What is the status of this pull request? Any pending changes, or is it blocked by anything?

**Comment by rmetzger** [comment]
Sorry for not reacting. I was busy with other stuff but I'll now address your comments.

**Comment by rmetzger** [comment]
I have those unit test style tests all in the KafkaITCase because they depend on the testing clusters started for the test.
They all need at least a running Zookeeper instance.
I can of course put them into a different class, but this will further slow down our tests because we spend more time starting and stopping zookeeper.

I've reworked the restore to reflect the open() / restoreState() order.
The PR has been updated.

**Comment by StephanEwen** [comment]
Looks good!

Will merge this..

**Pull request closed without merging**